src/PantheonTerminalWindow.vala

```diff
@@ -788,6 +802,97 @@ namespace PantheonTerminal {
             return (TerminalWidget)((Gtk.Bin)tab.page).get_child ();
         }
 
+        uint name_check_timeout_id = 0;
+        private void schedule_name_check () {
+            if (name_check_timeout_id > 0) {
+                Source.remove (name_check_timeout_id);
+            }
+
+            name_check_timeout_id = Timeout.add (50, () => {
+                if (!check_for_tabs_with_same_name ()) {
+                    return true;
+                } else {
+                    name_check_timeout_id = 0;
+                    return false;
+                }
+            });
+        }
+
+        /** Compare every tab label with every other and resolve ambiguities **/
+        private bool check_for_tabs_with_same_name () {
+            /* Take list copies so foreach clauses can be nested safely*/
+            var terms = terminals.copy ();
+            var terms2 = terminals.copy ();
+
+            foreach (TerminalWidget t in terms) {
+                string t_path = t.get_shell_location ();
+                string name = t.window_title;
+
+                if (name == "") { /* No point in continuing - tabs not finished updating */
+                    return false; /* Try again later */
+                }
+
+                /* Reset tab name to basename so long name only used when required */
+                t.tab_name = name;
+
+                foreach (TerminalWidget t2 in terms2) {
+                    string t2_path = t2.get_shell_location ();
+                    string t2_name = t2.window_title;
+
+                    if (t2 != t && t2_name == name) {
+                        if (t2_path != t_path) {
+                            t2.tab_name = disambiguate_name (name, t2_path, t_path);
+                            t.tab_name = disambiguate_name (name, t_path, t2_path);
+                            if (t == this.current_terminal) {
+                                this.title = t.tab_name;
+                            }
+                        }
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        private string disambiguate_name (string name, string path, string conflict_path) {
+            string prefix = "";
+            string prefix2 = "";
+            string pth = path;
+            string pth2 = conflict_path;
```

Comment:
Same question here. It seems like `conflict_path` is a better variable name than `pth2`